File: backend/app/services/estoque.py
from __future__ import annotations

from decimal import Decimal
from typing import Any, Iterable

_SIGN = {
    "compra":            Decimal("1"),
    "ajuste_positivo":   Decimal("1"),
    "saida_obra":        Decimal("-1"),
    "ajuste_negativo":   Decimal("-1"),
}


def saldo_from_movimentos(movimentos: Iterable[dict[str, Any]]) -> Decimal:
    total = Decimal("0")
    for m in movimentos:
        sign = _SIGN[m["tipo"]]
        total += sign * Decimal(str(m["quantidade"]))
    return total


def montar_saldos(
    materiais: list[dict[str, Any]],
    saldos_v: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    saldo_por_mat = {str(s["material_id"]): Decimal(str(s["saldo"])) for s in saldos_v}
    out: list[dict[str, Any]] = []
    for m in materiais:
        saldo = saldo_por_mat.get(str(m["id"]), Decimal("0"))
        minimo = Decimal(str(m.get("estoque_minimo") or 0))
        abaixo = minimo > 0 and saldo < minimo
        out.append({
            "material_id": m["id"],
            "sku": m["sku"],
            "nome": m["nome"],
            "categoria": m["categoria"],
            "unidade": m["unidade"],
            "saldo": saldo,
            "estoque_minimo": minimo,
            "abaixo_minimo": abaixo,
            "preco_unitario": Decimal(str(m["preco_unitario"])),
        })
    return out


def montar_analise_fabricacao(
    orcamento: dict[str, Any],
    produto: dict[str, Any],
    itens_orc: list[dict[str, Any]],
    saldos_v: list[dict[str, Any]],
    materiais: list[dict[str, Any]],
) -> dict[str, Any]:
    saldo_por_mat = {str(s["material_id"]): Decimal(str(s["saldo"])) for s in saldos_v}
    mat_index = {str(m["id"]): m for m in materiais}

    linhas: list[dict[str, Any]] = []
    faltantes = 0
    custo_reposicao = Decimal("0")
    for it in itens_orc:
        mat = mat_index.get(str(it["material_id"]), {})
        necessario = Decimal(str(it["quantidade"]))
        saldo_atual = saldo_por_mat.get(str(it["material_id"]), Decimal("0"))
        falta = max(Decimal("0"), necessario - saldo_atual)
        status = "faltante" if falta > 0 else "suficiente"
        preco = Decimal(str(mat.get("preco_unitario", it["preco_unitario"])))
        custo_linha = (falta * preco).quantize(Decimal("0.01"))
        if status == "faltante":
            faltantes += 1
            custo_reposicao += custo_linha
        linhas.append({
            "material_id": it["material_id"],
            "sku": mat.get("sku", ""),
            "nome": it["descricao"],
            "unidade": it["unidade"],
            "necessario": necessario,
            "saldo_atual": saldo_atual,
            "falta": falta,
            "status": status,
            "preco_unitario": preco,
            "custo_reposicao_linha": custo_linha,
        })

    return {
        "orcamento_id": orcamento["id"],
        "orcamento_numero": orcamento["numero"],
        "cliente_nome": orcamento["cliente_nome"],
        "produto_nome": produto["nome"],
        "itens": linhas,
        "totais": {
            "itens_total": len(linhas),
            "itens_faltantes": faltantes,
            "custo_reposicao": custo_reposicao,
        },
    }

File: frontend/src/lib/estoque.ts
export type MovimentoTipo =
  | 'compra'
  | 'ajuste_positivo'
  | 'saida_obra'
  | 'ajuste_negativo';

export interface Fornecedor {
  id: string;
  nome: string;
  cnpj?: string | null;
  contato_nome?: string | null;
  contato_email?: string | null;
  contato_fone?: string | null;
  observacao?: string | null;
  ativo: boolean;
  created_at: string;
  updated_at: string;
}

export interface FornecedorInput {
  nome: string;
  cnpj?: string | null;
  contato_nome?: string | null;
  contato_email?: string | null;
  contato_fone?: string | null;
  observacao?: string | null;
  ativo?: boolean;
}

export interface SaldoRow {
  material_id: string;
  sku: string;
  nome: string;
  categoria: string;
  unidade: string;
  saldo: string;            // backend sends Decimal as string via JSON
  estoque_minimo: string;
  abaixo_minimo: boolean;
  preco_unitario: string;
}

export interface Movimento {
  id: string;
  material_id: string;
  tipo: MovimentoTipo;
  quantidade: string;
  preco_unitario: string | null;
  fornecedor_id: string | null;
  orcamento_id: string | null;
  destino: string | null;
  nota_fiscal: string | null;
  observacao: string | null;
  criado_por: string;
  created_at: string;
}

export type MovimentoInput =
  | {
      tipo: 'compra';
      material_id: string;
      quantidade: string;
      preco_unitario: string;
      fornecedor_id: string;
      nota_fiscal?: string | null;
      observacao?: string | null;
    }
  | {
      tipo: 'saida_obra';
      material_id: string;
      quantidade: string;
      orcamento_id?: string | null;
      destino: string;
      observacao?: string | null;
    }
  | {
      tipo: 'ajuste_positivo' | 'ajuste_negativo';
      material_id: string;
      quantidade: string;
      observacao: string;
    };

export interface FabricacaoLinha {
  material_id: string;
  sku: string;
  nome: string;
  unidade: string;
  necessario: string;
  saldo_atual: string;
  falta: string;
  status: 'suficiente' | 'faltante';
  preco_unitario: string;
  custo_reposicao_linha: string;
}

export interface FabricacaoAnalise {
  orcamento_id: string;
  orcamento_numero: string;
  cliente_nome: string;
  produto_nome: string;
  itens: FabricacaoLinha[];
  totais: {
    itens_total: number;
    itens_faltantes: number;
    custo_reposicao: string;
  };
}

// ----- API calls -----

export type Fetcher = <T>(path: string, init?: RequestInit) => Promise<T>;

export const estoqueApi = {
  listSaldo: (
    fetchApi: Fetcher,
    opts: { abaixoMinimo?: boolean; q?: string } = {},
  ) => {
    const p = new URLSearchParams();
    if (opts.abaixoMinimo) p.set('abaixo_minimo', 'true');
    if (opts.q) p.set('q', opts.q);
    const qs = p.toString();
    return fetchApi<SaldoRow[]>(`/api/estoque/saldo${qs ? `?${qs}` : ''}`);
  },

  listMovimentos: (
    fetchApi: Fetcher,
    filters: Partial<{
      material_id: string;
      tipo: MovimentoTipo;
      fornecedor_id: string;
      orcamento_id: string;
      data_inicio: string;
      data_fim: string;
      limit: number;
      offset: number;
    }> = {},
  ) => {
    const p = new URLSearchParams();
    for (const [k, v] of Object.entries(filters)) {
      if (v !== undefined && v !== null && v !== '') p.set(k, String(v));
    }
    const qs = p.toString();
    return fetchApi<Movimento[]>(`/api/estoque/movimento${qs ? `?${qs}` : ''}`);
  },

  createMovimento: (fetchApi: Fetcher, body: MovimentoInput) =>
    fetchApi<Movimento>('/api/estoque/movimento', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  analiseFabricacao: (fetchApi: Fetcher, orcamentoId: string) =>
    fetchApi<FabricacaoAnalise>(`/api/estoque/fabricacao/${orcamentoId}`),
};

export const fornecedorApi = {
  list: (fetchApi: Fetcher, opts: { ativo?: boolean; q?: string } = {}) => {
    const p = new URLSearchParams();
    if (opts.ativo !== undefined) p.set('ativo', String(opts.ativo));
    if (opts.q) p.set('q', opts.q);
    const qs = p.toString();
    return fetchApi<Fornecedor[]>(`/api/fornecedor${qs ? `?${qs}` : ''}`);
  },

  create: (fetchApi: Fetcher, body: FornecedorInput) =>
    fetchApi<Fornecedor>('/api/fornecedor', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  update: (fetchApi: Fetcher, id: string, body: Partial<FornecedorInput>) =>
    fetchApi<Fornecedor>(`/api/fornecedor/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  deactivate: (fetchApi: Fetcher, id: string) =>
    fetchApi<{ ok: boolean }>(`/api/fornecedor/${id}`, {
      method: 'DELETE',
    }),
};
